base/shared/src/main/scala/scalaz/data/acatenable1.scala

```diff
@@ -51,14 +48,31 @@ sealed abstract class ACatenable1[=>:[_, _], A, B] {
 
   /**
    * Compose the leafs in a balanced binary fashion.
+   *
+   * Implementation is easiest to understand in terms of `AList1.foldBalanced`;
+   * first, we reassociate all left-nested trees to be right-nested. Second,
+   * we do exactly what `AList1.foldBalanced` does by folding from the left
+   * with `PostComposeBalancer`, because a right-nested `ACatenable1` *is*
+   * an `AList1`.
    */
   @tailrec
-  final def fold(implicit ev: Semicategory[=>:]): A =>: B = this match {
-    case Chain(Chain(f, g), h) => (f >>> (g >>> h)).fold
+  final def foldBalanced(implicit ev: Semicategory[=>:]): A =>: B = this match {
+    case Chain(Chain(f, g), h) => (f :++ (g :++ h)).foldBalanced
     case Chain(Lift(f), g) =>
       g.foldLeft[PostComposeBalancer[=>:, A, ?]](PostComposeBalancer(f))(PostComposeBalancer.rightAction).result
     case Lift(f) => f
   }
+
+  /**
+   * Allows you to view the internal structure of an ACatenable1.
+   * Not stack-safe.
+   */
+  final def fold[G[_, _]](f: Forall3[λ[(X, Y, Z) => (G[X, Y], G[Y, Z]) => G[X, Z]]])(z: =>: ~~> G): G[A, B] =
```

Comment:
I suppose so, but why would anyone use that? It's not going to be stack-safe. Perhaps for performance?

Reply:
Okay, maybe not exactly like this one—it should at least be stack safe—but otherwise it would do the most efficient folding of `Alist1` (`foldLeft`?). It could be used when it is safe to use non-balanced composition in the target semicategory `G`.

Reply:
Afaik the fastest must not be stack-safe; associativity gives us the structure to process as a sort of tail-recursive stack machine. Otherwise, we're going to have to use fully-general trampolining.

`fold` as it's presented here though remains the fastest.

Reply:
Yeah, I meant the most efficient among stack-safe implementations.

Actually, I don't even mind if it internally just calls `foldBalanced`. My (very minor) concern was only that when the user doesn't care about balanced, calling `foldBalanced` could feel to her like she is over-stating her needs.

Reply:
True, the thing is I kind of like forcing the user to explicitly make a choice here.

Reply:
I mean when the choice of folding strategy _really_ doesn't make any difference for the user. Trivial example: `type G[X, Y] = Int` and composition is addition. Then you would be forcing the user to make an arbitrary choice, when she would rather leave the freedom to the implementation to choose whatever works best for the data structure.

Reply:
I guess that's fair; I'll add these shortly.